test/ExecuteUpdateTest.cpp

```diff
@@ -0,0 +1,259 @@
+              testing::Eq(VocabIndex(1)));
+  EXPECT_THAT(ExecuteUpdate::resolveVariable(idTable, 1, 1UL),
+              testing::Eq(VocabIndex(4)));
+  EXPECT_THAT(ExecuteUpdate::resolveVariable(idTable, 2, 1UL),
+              testing::Eq(std::nullopt));
+  EXPECT_THAT(ExecuteUpdate::resolveVariable(idTable, 2, Id::makeUndefined()),
+              testing::Eq(std::nullopt));
+}
+
+TEST(ExecuteUpdate, computeAndAddQuadsForResultRow) {
+  auto VocabIndex = [](const uint64_t index) {
+    return Id::makeFromVocabIndex(VocabIndex::make(index));
+  };
+  auto idTable = makeIdTableFromVector(
+      {{VocabIndex(0), VocabIndex(1), VocabIndex(2)},
+       {VocabIndex(3), VocabIndex(4), VocabIndex(5)},
+       {VocabIndex(6), Id::makeUndefined(), VocabIndex(8)}});
+  auto expectComputeQuads =
+      [](const std::vector<ExecuteUpdate::TransformedTriple>& templates,
+         const IdTable& idTable, uint64_t rowIdx,
+         const testing::Matcher<const std::vector<IdTriple<>>&>&
+             expectedQuads) {
+        std::vector<IdTriple<>> result;
+        ExecuteUpdate::computeAndAddQuadsForResultRow(templates, result,
+                                                      idTable, rowIdx);
+        EXPECT_THAT(result, expectedQuads);
+      };
+  // Compute the quads for an empty template set yields no quads.
+  expectComputeQuads({}, idTable, 0, testing::IsEmpty());
+  // Compute the quads for template without variables yields the templates
+  // unmodified.
+  expectComputeQuads(
+      {{VocabIndex(0), VocabIndex(1), VocabIndex(2), VocabIndex(3)}}, idTable,
+      0,
+      testing::ElementsAreArray({IdTriple{
+          {VocabIndex(0), VocabIndex(1), VocabIndex(2), VocabIndex(3)}}}));
+  expectComputeQuads(
+      {{VocabIndex(0), VocabIndex(1), VocabIndex(2), VocabIndex(3)}}, idTable,
+      1,
+      testing::ElementsAreArray({IdTriple{
+          {VocabIndex(0), VocabIndex(1), VocabIndex(2), VocabIndex(3)}}}));
+  // The variables in templates are resolved to the value of the variable in the
+  // specified row of the result.
+  expectComputeQuads(
+      {{0UL, VocabIndex(1), 1UL, VocabIndex(3)}}, idTable, 0,
+      testing::ElementsAreArray({IdTriple{
+          {VocabIndex(0), VocabIndex(1), VocabIndex(1), VocabIndex(3)}}}));
+  expectComputeQuads(
+      {{0UL, VocabIndex(1), 1UL, VocabIndex(3)}}, idTable, 1,
+      testing::ElementsAreArray({IdTriple{
+          {VocabIndex(3), VocabIndex(1), VocabIndex(4), VocabIndex(3)}}}));
+  // Quads with undefined IDs cannot be stored and are not returned.
+  expectComputeQuads({{0UL, VocabIndex(1), 1UL, VocabIndex(3)}}, idTable, 2,
+                     testing::IsEmpty());
+  expectComputeQuads(
+      {{VocabIndex(0), VocabIndex(1), Id::makeUndefined(), VocabIndex(3)}},
+      idTable, 0, testing::IsEmpty());
+  // Some extra cases to cover all branches.
+  expectComputeQuads(
+      {{Id::makeUndefined(), VocabIndex(1), VocabIndex(2), VocabIndex(3)}},
+      idTable, 0, testing::IsEmpty());
```

Comment:
also shorten the `VocabIndex`(see above), and maybe you can make the `IsEmpty` a default argument for the localVocabMather argument.

Reply:
That's a matcher for the computed quads. I'm not convinced that hiding the only matcher in this case is a good idea, because empty imo is not a intuitive default. The high number of tests with empty results stems from the edge cases where quads might not be returned. Maybe as an alternative add a function `expectComputeQuadsIsEmpty`?